KNX Component: Scene support and expose sensor values

Julius2342 opened the pull request:
## Description:

* Bumped version of KNX library to 0.8.0
* Support for many more sensor types: power, electric_current, electriy_potential, energy, frequency, heatflowrate, phaseanglerad, phaseangledeg, powerfactor.
* Added scene support.
* Added support for exposing sensors to KNX bus 
* Added reset function for binary sensors.

## Broadcast sensor values and time to KNX bus

This feature was discussed within PR #10708 and PR #10654 and is now a more generic approach. 

In the new implementation, all changes of sensor values will be broadcasted to KNX bus. Additionally HASS will answer read requests from KNX bus.

```yaml
# Example configuration.yaml entry
knx:
    expose:
        - type: 'temperature'
          entity_id: 'sensor.owm_temperature'
          address: '0/0/2'
        - type: 'time'
          address: '0/0/1'
        - type: 'datetime'
          address: '0/0/23'
```

For available sensor types see [KNX sensor documentation](https://home-assistant.io/components/sensor.knx/).

## Scene support

```yaml
# Example configuration.yaml entry
scene:
  - name: Romantic
    platform: knx
    address: 8/8/8
    scene_number: 23
```

## reset_after within binary sensors

Added new parameter `reset_after` to configuration. Binary sensor will reset state after given milliseconds. This configuration is helpful for touch buttons which don't send an off-telegram after being released.

```yaml
binary_sensor:
- platform: knx
  name: ToiletButton
  address: '5/2/0'
  reset_after: 100
```

## home-assistant.github.io

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>



## Checklist:
  - [X] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io): [PR](https://github.com/home-assistant/home-assistant.github.io/pull/4575/files)

If the code communicates with devices, web services, or third-party tools:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

Julius2342 committed:
XKNX improvements: Added Scene support, added support for exposing sensors to KNX bus, added reset value option for binary switches

fabaff commented:
```bash
E:305, 8: No name 'datetime' in module 'xknx.devices' (no-name-in-module)
E:305, 8: Unable to import 'xknx.devices.datetime' (invalid syntax (<string>, line 59)) (import-error)
```

Julius2342 committed:
fixed import

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into xknx

Julius2342 commented:
@fabaff : i fixed the pylint problem (had to wait for the next release of the underlaying library).

Julius2342 committed:
Bumped version of KNX library (minor upgrade with two important bugfixes)

Julius2342 committed:
bumped version of xknx (now without python requirement *sigh*)

balloob commented:
Please try to keep 1 feature per PR

balloob commented on homeassistant/components/knx.py:
```diff
@@ -203,6 +221,24 @@ def register_callbacks(self):
+    def create_exposures(self):
+        """Create exposures."""
+        exposures = []
+        if CONF_KNX_EXPOSE in self.config[DOMAIN]:
```
guard clause.

```
if CONF_KNX_EXPOSE not in …:
    return exposures
```

balloob commented on homeassistant/components/scene/knx.py:
```diff
@@ -0,0 +1,95 @@
+def async_setup_platform(hass, config, async_add_devices,
+                         discovery_info=None):
+    """Set up the scenes for KNX platform."""
+    if DATA_KNX not in hass.data \
```
This is not possible because you have `knx` as a dependency and so setup never gets called unless knx has been setup correctly.

Also, just returning `False` is weird. Then it will never be setup?

balloob commented on homeassistant/components/scene/knx.py:
```diff
@@ -0,0 +1,95 @@
+
+    def activate(self):
+        """Activate the scene."""
+        self.hass.async_add_job(self.scene.run())
```
Not allowed to call async methods in sync context. Also don't run outside of the main thread. Just call `self.scene.run()`

balloob commented on homeassistant/components/scene/knx.py:
```diff
@@ -0,0 +1,95 @@
+        return self.scene.name
+
+    @property
+    def should_poll(self):
```
Default for scenes is already false.

balloob commented on homeassistant/components/scene/knx.py:
```diff
@@ -0,0 +1,95 @@
+        return False
+
+    @property
+    def is_on(self):
```
This doesn't exist.

balloob commented on homeassistant/components/knx.py:
```diff
@@ -249,3 +285,63 @@ def __init__(self, hass, device, hook, action, counter=1):
+            broadcast_type=broadcast_type,
+            group_address=self.address)
+        self.xknx.devices.add(self.device)
+        async_track_utc_time_change(
```
This is so weird. Shouldn't this be a feature inside xknx ?

Julius2342 committed:
Issue #11978: fixed review comments

Julius2342 committed:
Issue #11978: hound suggestion fixed:

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into xknx

MartinHjelmare commented on homeassistant/components/scene/knx.py:
```diff
@@ -0,0 +1,82 @@
+
+    def __init__(self, hass, scene):
+        """Init KNX scene."""
+        self.hass = hass
```
This isn't used.

MartinHjelmare commented on homeassistant/components/scene/knx.py:
```diff
@@ -0,0 +1,82 @@
+    else:
+        async_add_devices_config(hass, config, async_add_devices)
+
+    return True
```
Nothing is checking this return value. Please remove this.

MartinHjelmare commented on homeassistant/components/knx.py:
```diff
@@ -249,3 +285,61 @@ def __init__(self, hass, device, hook, action, counter=1):
+
+    def __init__(self, hass, xknx, expose_type, address):
+        """Initialize of Expose class."""
+        self.hass = hass
```
This doesn't seem to be used.

Julius2342 committed:
review comments

Julius2342 commented:
@balloob @MartinHjelmare : I think I addressed all mentioned issues.

[In case you wonder why i just removed the async_track_utc_time_change without bumping the version of the lib. xknx already sends time regularly when a DateTime object is added to the internal devices list. This does not happen every minute but every hour, which should be ok for KNX devices. ]

balloob commented on homeassistant/components/knx.py:
```diff
@@ -249,3 +285,60 @@ def __init__(self, hass, device, hook, action, counter=1):
+            group_address=self.address,
+            value_type=self.type)
+        self.xknx.devices.add(self.device)
+        async_track_state_change(
```
This is an async method. That means that this method is only allowed to be called from an async context. This doesn't seem to be the case.

Julius2342 committed:
Merge branch 'dev' into xknx

Julius2342 committed:
made async functions async

Julius2342 commented:
@balloob : is 672a34f what you had in mind?

MartinHjelmare commented on homeassistant/components/binary_sensor/knx.py:
```diff
@@ -49,6 +50,7 @@
     vol.Optional(CONF_DEVICE_CLASS): cv.string,
     vol.Optional(CONF_SIGNIFICANT_BIT, default=CONF_DEFAULT_SIGNIFICANT_BIT):
         cv.positive_int,
+    vol.Optional(CONF_RESET_AFTER, default=None): cv.positive_int,
```
Remove default `None`. Default value must pass validation in latest voluptuous version.

Julius2342 replied:
will change it to:

```
+    vol.Optional(CONF_RESET_AFTER, default=None): vol.Any(
+        None, cv.positive_int),
```

MartinHjelmare commented on homeassistant/components/knx.py:
```diff
@@ -26,6 +28,10 @@
 CONF_KNX_STATE_UPDATER = "state_updater"
+CONF_KNX_EXPOSE = "expose"
+CONF_KNX_EXPOSE_TYPE = "type"
+CONF_KNX_EXPOSE_ENTITY_ID = "entity_id"
```
Maybe use the existing entity_id constant from const.py?

Julius2342 replied:
fixed

MartinHjelmare commented on homeassistant/components/knx.py:
```diff
@@ -45,6 +51,12 @@
 })
 
+EXPOSE_SCHEMA = vol.Schema({
+    vol.Required(CONF_KNX_EXPOSE_TYPE): cv.string,
```
Do you know these possible types already? If so validate that the type is one of them.

Julius2342 replied:
well, it is a growing list of 30 different types (KNX itself has about 100 different types). And same list as it is used within `sensor/knx.py`. I tend to say maintaining this list on 3 different places is too much overhead and is error-prone.

Or is there any possibility to us an external lib within validation?

MartinHjelmare replied:
Requirements are not installed at the time of validation, last time I looked. So 3rd party library usage during validation is not possible.

Ok. Just curious what's the third place to maintain a list? I understand python knx library and homeassistant are two.

Julius2342 replied:
https://github.com/XKNX/xknx/blob/master/xknx/devices/remote_value_sensor.py#L22

MartinHjelmare commented on homeassistant/components/knx.py:
```diff
@@ -248,3 +287,60 @@ def __init__(self, hass, device, hook, action, counter=1):
+    def async_register(self):
+        """Register listener."""
+        from xknx.devices import DateTime, DateTimeBroadcastType
+        broadcast_type_string = self.type.upper()
```
Can this be done in the validation schema? Or do you need type to be different from broadcast_type?

Julius2342 replied:
`self.type` is used within `KNXExposeTime` to look up the right entry within `DateTimeBroadcastType`  -Enum and within `ExposeSensor` which needs a lower case value.

MartinHjelmare commented on homeassistant/components/scene/knx.py:
```diff
@@ -0,0 +1,79 @@
+
+@callback
+def async_add_devices_config(hass, config, async_add_devices):
+    """Set up scene for KNX platform configured within plattform."""
```
Typo platform.

Julius2342 replied:
fixed

Julius2342 committed:
Addressed issues mentined by @MartinHjelmare

MartinHjelmare commented on homeassistant/components/binary_sensor/knx.py:
```diff
@@ -49,6 +50,8 @@
     vol.Optional(CONF_DEVICE_CLASS): cv.string,
     vol.Optional(CONF_SIGNIFICANT_BIT, default=CONF_DEFAULT_SIGNIFICANT_BIT):
         cv.positive_int,
+    vol.Optional(CONF_RESET_AFTER, default=None): vol.Any(
```
I'd prefer if you remove the default. You are using `dict.get` below so the default here has no effect.

Julius2342 replied:
changed it to: `vol.Optional(CONF_RESET_AFTER): cv.positive_int,`

MartinHjelmare commented on homeassistant/components/knx.py:
```diff
@@ -45,6 +51,12 @@
 
+EXPOSE_SCHEMA = vol.Schema({
+    vol.Required(CONF_KNX_EXPOSE_TYPE): cv.string,
+    vol.Optional(ATTR_ENTITY_ID): cv.string,
```
Use `CONF_ENTITY_ID` as this is for a config value in a config schema.

Julius2342 committed:
removed default=None from validation schema

Julius2342 commented:
@MartinHjelmare : Is there anything missing from your perspective? Are you fine with my comment regarding `upper()` or - if not - do you have any idea how i can solve this?

MartinHjelmare commented:
That's fine but the comment about `CONF_ENTITY_ID` should be taken care of.

Julius2342 committed:
ATTR_ENTITY_ID->CONF_ENTITY_ID

Julius2342 commented:
@MartinHjelmare : Oh ***blush***! oversaw this comment! Pushed with commit 35339d1341fe56981839d0217840d944e7fbef78

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into xknx

Julius2342 committed:
Merge branch 'dev' into xknx

Julius2342 committed:
moved missing function to async syntax

Julius2342 committed:
pylint

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into xknx

Julius2342 committed:
Trigger notification

Julius2342 committed:
Trigger notification

Julius2342 commented:
@MartinHjelmare : May I ask you to have a look?

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented on homeassistant/components/knx.py:
```diff
@@ -45,6 +51,12 @@
 
+EXPOSE_SCHEMA = vol.Schema({
+    vol.Required(CONF_KNX_EXPOSE_TYPE): cv.string,
+    vol.Optional(CONF_ENTITY_ID): cv.string,
```
This would be even better validated with `cv.entity_id`.

MartinHjelmare reviewed (commented):
Sorry, missed one.

Julius2342 committed:
fixed review comment

MartinHjelmare reviewed (approved):
Great!
